Feature/ca certs dir

aronerben opened the pull request:
https://github.com/oxidizing/letters/issues/10

aronerben committed:
Add ca-certs package

To assist fetching the appropriate CA certificate file
paths, use ca-certs package.

aronerben committed:
Make config a module

To facilitate config building and hiding the type implementation, make
it a module with builder functions for optionals.

aronerben committed:
Detach sender from config

To simplify the transport config, remove the sender mail from the
config. It has to be provided when building the mail and sending
it. Additionally, improve variable naming.

aronerben committed:
Detect ca dirs

To handle the case where no ca dir is provided by the user, use
ca-certs to find possible locations based on the OS.

aronerben committed:
Improve error message

Improve missing CA cert directory error message.

aronerben committed:
Add ca-certs to ci

To run tests, ca-certs package has to be added to ci.

aronerben committed:
Replace ca-certs package with local implementation

Because we always pinning opam packages is not a sustainable solution,
add a local implementation of the ca-certs package.

aronerben committed:
Merge branch 'master' into feature/ca-certs-dir

mikonieminen commented on lib/letters.ml:
```diff
@@ -205,8 +214,15 @@ let send ~config:c ~recipients:r ~message:m =
   let* certs =
-    match Fpath.of_string c.ca_dir with
+    match Fpath.of_string path with
     | Ok path -> certs_of_pem_directory ~ext:"pem" path
```
Does this really work? I'm a bit puzzled what the `path` ends up being and how this `certs_of_pem_directory` works when our detected files are actually CA bundles and not the CA path while here we only read in all `pem` files.

aronerben replied:
No, you're right. This won't work with the automatic CA detection. I will rewrite it to use bundles, because the library will probably provide that.

mikonieminen commented on lib/letters.mli:
```diff
@@ -1,27 +1,43 @@
+   ** [username] username needed for the login
+   ** [password] user's password for the login
+   ** [hostname] hostname of the SMTP server
+   ** *)
```
Documentation for `with_starttls` is now missing.

aronerben replied:
Thanks, I messed up copy pasting, also added `sender` accidentally...

aronerben committed:
Fix documentation for configuration builder

aronerben committed:
Fix formatting

To have consistent formatting, run ocamlformatter.

aronerben committed:
Make ca-certs finding work with bundles

To be able to use ca-certs library, ca certs have to be given as a
bundle and not as a directory. Adjust record namings accordingly.

aronerben commented:
@mikonieminen What is the status of this PR? Can I merge it?

mikonieminen commented:
Can you rebase your change and clean-up commits. Now the history is a bit messy.

aronerben commented:
Closed due to new PR: https://github.com/oxidizing/letters/pull/23
